File: 234-palindrome-linked-list/234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* findMiddle (ListNode *head) {
        ListNode *slow = head;
        ListNode *fast = head;
        
        while (fast->next != NULL && fast->next->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
    
    ListNode* reverse(ListNode* head) {
        
        // create previous and current node
        
        ListNode *prevNode = NULL;
        ListNode *currentNode = head;
        
        while (currentNode != NULL) {
            ListNode *nextNode = currentNode->next;
            currentNode->next = prevNode;
            prevNode = currentNode;
            currentNode = nextNode;
        }
        return prevNode;
    }
    
    
    bool isPalindrome(ListNode* head) {
        
        // find middle node
        
        ListNode *middle = findMiddle(head);
        
        // reverse the list from middle
        
        ListNode *secondHalf = reverse(middle->next);
        
        // compare each value from first half and second half
        
        ListNode *firstHalf = head;
        while (secondHalf != NULL) {
            if (firstHalf->val != secondHalf->val) return false;
            firstHalf = firstHalf->next;
            secondHalf = secondHalf->next;
        }
        return true;
    }
};

File: 234-palindrome-linked-list/234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    ListNode findMiddle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
    
    ListNode reverseSecondHalf(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        
        while (curr != null) {
            ListNode nextNode = curr.next;
            
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }
        return prev;
        
    }
    
    public boolean isPalindrome(ListNode head) {
        
        if (head == null || head.next == null) return true;
        
        ListNode middleNode = findMiddle(head);
        ListNode secondHalfStart = reverseSecondHalf(middleNode.next);
        
        ListNode firstHalf = head;
        
        while (secondHalfStart != null) {
            if (firstHalf.val != secondHalfStart.val) {
                return false;
            }
            firstHalf = firstHalf.next;
            secondHalfStart = secondHalfStart.next;
        }
        return true;    
    }
}